File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.py
def solution(keymap, targets):
    answer = []
    keydict = {}
    
    for key in keymap:
        for i, k in enumerate(key):
            keydict[k] = min(i + 1, keydict[k]) if k in keydict else i + 1
    
    for target in targets:
        count = 0
        for t in target:
            if t not in keydict:
                count = -1
                break
            count += keydict[t]
        answer.append(count)
    return answer

File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.java
import java.util.HashMap;

class Solution {
    public int[] solution(String[] keymap, String[] targets) {
        int[] answer = new int[targets.length] ;
        
        HashMap<Character, Integer> min_index = new HashMap<>();
        
        for (String key : keymap) {
            for (int i = 0; i < key.length(); i++) {
                char ch = key.charAt(i); 
                if (!min_index.containsKey(ch) || i < min_index.get(ch)) {
                    min_index.put(ch, i + 1);
                }
            }
        }
        
        
        for (int i = 0; i < targets.length; i++) {
            char[] arr = targets[i].toCharArray();
            for(char ch : arr) {
                if(min_index.containsKey(ch)) {
                    answer[i] += min_index.get(ch);
                }
                else {
                    answer[i] = -1;
                    break;
                }
            }
        } 
        
        return answer;
    }
}